bitcoin/src/crypto/key.rs

```diff
@@ -231,12 +229,26 @@ impl fmt::Display for PublicKey {
 }
 
 impl FromStr for PublicKey {
-    type Err = Error;
-    fn from_str(s: &str) -> Result<PublicKey, Error> {
+    type Err = ParsePublicKeyError;
+    fn from_str(s: &str) -> Result<PublicKey, ParsePublicKeyError> {
         match s.len() {
-            66 => PublicKey::from_slice(&<[u8; 33]>::from_hex(s)?),
-            130 => PublicKey::from_slice(&<[u8; 65]>::from_hex(s)?),
-            len => Err(Error::InvalidHexLength(len)),
+            66 => {
+                PublicKey::from_slice(&<[u8; 33]>::from_hex(s).map_err(|op| {
+                    match op {
+                        HexToArrayError::Conversion(HexToBytesError::InvalidChar(char)) => ParsePublicKeyError::InvalidChar(char),
+                        _ => panic!("This should not have happened"),
```

Comment:
Me too but really it's not a huge deal and this PR was around for a long time...

Reply:
> this PR was around for a long time

I think new contributors should be held to a very high standard on their first few PRs. @Harshit933 is doing a good job so far, I say let him polish it, then he has a PR he can be proud of. (Yes I know re-reviewing is a PITA.)

Reply:
Sure, no problem but I'm keeping my ACK. :)

Reply:
The code would be like this :
```rust
match s.len() {
            66 => {
                PublicKey::from_slice(&<[u8; 33]>::from_hex(s).map_err(|op| {
                    match op {
                        HexToArrayError::Conversion(HexToBytesError::InvalidChar(char)) => ParsePublicKeyError::InvalidChar(char),
                        HexToArrayError::InvalidLength(_,_) => unreachable!("Length is already pre checked"),
                        HexToArrayError::Conversion(HexToBytesError::OddLengthString(_)) => unreachable!("Length is already pre checked and it can't be odd")
                    }
                })?).map_err(From::from)
            },
            130 => {
                PublicKey::from_slice(&<[u8; 65]>::from_hex(s).map_err(|op| {
                    match op {
                        HexToArrayError::Conversion(HexToBytesError::InvalidChar(char)) => ParsePublicKeyError::InvalidChar(char),
                        HexToArrayError::InvalidLength(_,_) => unreachable!("we checked the length"),
                        HexToArrayError::Conversion(HexToBytesError::OddLengthString(_)) => unreachable!("we checked the length")
                    }
                })?).map_err(From::from)
            }
            len => Err(ParsePublicKeyError::InvalidHexLength(len)),
        }
```
In this the problem is the code is being repeated. Thus, it's kind of bad. For now I can only think of a solution of making an helper function that just returns the error for me. But, I have noticed the code follows some kind of pattern (I don't know the pattern) such that if I put a helper function it would break the pattern. What can I do here?

Reply:
I think you can let it repeat. A bit of repetition is not bad, we're not breaking single source of truth here and I hope we can improve hex API to eventually remove the match entirely.

Reply:
Perhaps this is cleaner:
```rust
match s.len() {
    66 => {
        PublicKey::from_slice(&<[u8; 33]>::from_hex(s).map_err(|op| {
            match op {
                Conversion(HexToBytesError::InvalidChar(char)) => ParsePublicKeyError::InvalidChar(char),
                Conversion(HexToBytesError::OddLengthString(_)) | InvalidLength(_,_) => unreachable!("valid length"),
            }
        })?).map_err(From::from)
    },
    130 => {
        PublicKey::from_slice(&<[u8; 65]>::from_hex(s).map_err(|op| {
            match op {
                Conversion(HexToBytesError::InvalidChar(char)) => ParsePublicKeyError::InvalidChar(char),
                Conversion(HexToBytesError::OddLengthString(_)) | InvalidLength(_,_) => unreachable!("valid length"),
            }
        })?).map_err(From::from)
    }
    len => Err(ParsePublicKeyError::InvalidHexLength(len)),
}
```

Reply:
Thanks, I will rush to university now.